README.rst
#######
Redisdb
#######

.. image:: https://pypip.in/wheel/django-redisdb/badge.svg
    :target: https://pypi.python.org/pypi/django-redisdb/
    :alt: Wheel Status

.. image:: https://pypip.in/version/django-redisdb/badge.svg
    :target: https://pypi.python.org/pypi/django-redisdb/
    :alt: Latest Version

.. image:: https://pypip.in/license/django-redisdb/badge.svg
    :target: https://pypi.python.org/pypi/django-redisdb/
    :alt: License

.. image:: https://travis-ci.org/kidosoft/django-redisdb.svg?branch=master
    :target: https://travis-ci.org/kidosoft/django-redisdb
    :alt: Build status

.. image:: https://coveralls.io/repos/kidosoft/django-redisdb/badge.svg
    :target: https://coveralls.io/r/kidosoft/django-redisdb
    :alt: Coverage

.. image:: https://readthedocs.org/projects/django-redisdb/badge/?format=svg
    :target: https://django-redisdb.readthedocs.org
    :alt: Documetation


Django-redisdb is Redis backend for Django that allows 
using Redis as a cache and as a database at the same time.
Django-redisdb provides backends for master/master and sharded configuration.

Installation
============

.. code-block:: console

   pip install django-redisdb

Quick usage guide
=================


In settings.py:

.. code-block:: python

    CACHES = {
        'redis_ring': {
            'BACKEND': 'redisdb.backends.RedisRing',  # sharding backend
            'DB': 0,
            'LOCATION': [
                'localhost:6379',
                'localhost:6380',
            ],
        },
        'redis_copy': {
            'BACKEND': 'redisdb.backends.RedisCopy',  # copying backend
            'DB': 0,
            'LOCATION': [
                'localhost:6379',
                'localhost:6380',
            ],
        }
    }

Usage:

.. code-block:: python

   >>> from django.core.cache import caches
   >>> caches['redis_ring'].set('one_key', 123)  # set key1 only on on server
   [True]
   >>> caches['redis_copy'].set('other_key', 234)  # set key2 on all servers
   [True, True]

With RedisRing value is set only on one node. With RedisCopy it's set on all
nodes (two nodes in examle above).

Redis is much more powerfull then simple cache. It should be seen
as a specialized database. With django-redisdb you can use all its power.
For example you can use redis' sorted sets:

.. code-block:: python

    >>> caches['redis_copy'].zadd('myzset', 1, 'one')
    [0, 1]
    >>> caches['redis_copy'].zadd('myzset', 2, 'two')
    [0, 1]
    >>> caches['redis_copy'].zadd('myzset', 3, 'three')
    [0, 1]
    >>> caches['redis_copy'].zrange('myzset', 0, -1)
    ['one', 'two', 'three']
    >>> caches['redis_copy'].zrange('myzset', 0, -1, withscores=True)
    [('one', 1.0), ('two', 2.0), ('three', 3.0)]


Supported Django versions
=========================

django-redisdb runs on Django 1.2 to Django 1.8

Documentation
=============

Full documentation is available at http://django-redisdb/en/latest/index.html


CHANGES.rst
Version History
===============================================================================

Version: 0.2.0 (Unreleased)

ADDED
^^^^^

  * support for configuration options


Version: 0.1.0 (2017-12-01)
-------------------------------------------------------------------------------

ADDED
^^^^^

  * initial release


pavement.py
# -*- coding: utf-8 -*-
import glob
import os
import sys

from paver.easy import *  # noqa
try:
    import paver.doctools  # noqa
except ImportError:
    pass
sys.path.append('.')

options(
    project=Bunch(
        name='redisdb',
        package_name='django-redisdb',
    ),
    sphinx=Bunch(
        builddir="_build",
        apidir=None,
    ),
    coverage=Bunch(
        omit=[
            'pavement.py',
            'manage.py',
            'setup.py',
            'example/*',
        ]
    ),
)


@task
def cleanup():
    """ Removes generated files. """
    sh('rm -rf build', ignore_error=True)
    sh('rm -rf dist', ignore_error=True)
    sh('rm -rf */*.egg-info', ignore_error=True)
    sh('rm -rf htmlcov', ignore_error=True)
    sh('rm -rf docs/_build', ignore_error=True)
    sh('rm .coverage.*', ignore_error=True)
    sh("find . -name '*.pyc' -delete", ignore_error=True)
    sh("find . -name '__pycache__' -delete", ignore_error=True)


@task
def kwalitee(options):
    """ Check for kwalitee. """
    sh('flake8 src')
    sh('pep257 src', ignore_error=True)


@task
def sdist():
    """ Generate source distribution. """
    sh('python setup.py sdist -q')


@task
def bdist_wheel():
    """ Generate binary distribution. """
    sh('python setup.py bdist_wheel --universal -q')


@task
@needs('sdist', 'bdist_wheel')
def build():
    """ Build packages. """
    pass


@task
@needs('build')
def test_install(options):
    """ Test installation of packages. """
    distfiles = glob.glob('dist/*')
    venv_bin_dir = '.tox/status/bin'
    pip_path = os.path.join(venv_bin_dir, 'pip')
    python_path = os.path.join(venv_bin_dir, 'python')
    for idx, distfile in enumerate(distfiles):
        try:
            # install package
            sh('%s install -q %s' % (pip_path, distfile))
            # test package installation
            sh('%s -c "import %s"' % (python_path, options.name))
        finally:
            # uninstall
            sh('%s uninstall -q -y %s' % (pip_path, options.package_name), ignore_error=True)


@task
def test_all(options):
    """ Run tests in different environtemtns. """
    sh('tox')


@task
@needs('cleanup', 'kwalitee', 'test_all', 'test_install', 'html')
def pre_release(options):
    """ Check project before release. """
    pass


@task
@cmdopts([
    ("path=", "p", "Docs path"),
])
@needs('cleanup', 'html')
def publish_docs(options):
    """ Uploads docs to server. """
    path = options.get('path', None)
    if path is None:
        path = os.environ.get('DOCS_PATH', '')
    sh('''sed -i '' 's/href="\(http:\/\/sphinx-doc.org\)/rel="nofollow" href="\\1"/' docs/_build/html/*.html''')
    sh('rsync -av docs/_build/html/ %s/%s/' % (path, options.package_name))


@task
def sign_dist(options):
    for distfile in glob.glob('dist/*'):
        if distfile.endswith('.tar.gz') or distfile.endswith('.whl'):
            sh('gpg --detach-sign -a %s' % distfile)


@task
def twine_upload(options):
    sh('twine upload dist/*')


@task
@needs('cleanup', 'build', 'sign_dist', 'twine_upload')
def release(options):
    """ Generate packages and upload to PyPI. """
    pass


example/settings.py
"""
Django settings for example project.

For more information on this file, see
https://docs.djangoproject.com/en/1.7/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/1.7/ref/settings/
"""
# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
import os
BASE_DIR = os.path.dirname(os.path.dirname(__file__))


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/1.7/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = '4-aol-=3jk=^d%ai$xf95gpx^)0@0tf&g5(0uxp*d#-%7**^u%'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

TEMPLATE_DEBUG = True

ALLOWED_HOSTS = []


# Application definition

INSTALLED_APPS = (
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'redisdb',
)

MIDDLEWARE_CLASSES = (
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
)

ROOT_URLCONF = 'example.urls'

WSGI_APPLICATION = 'example.wsgi.application'


# Database
# https://docs.djangoproject.com/en/1.7/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}

# Internationalization
# https://docs.djangoproject.com/en/1.7/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.7/howto/static-files/

STATIC_URL = '/static/'
LOGIN_URL = '/admin/'

CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.dummy.DummyCache',
    },
    'redis_ring': {
        'BACKEND': 'redisdb.backends.RedisRing',
        'DB': '0',
        'LOCATION': [
            'localhost:6379',
            'localhost:6380',
        ]
    },
    'redis_copy': {
        'BACKEND': 'redisdb.backends.RedisCopy',
        'DB': '0',
        'LOCATION': [
            'localhost:6379',
            'localhost:6380',
        ]
    },
}
